// @jasonfleischer/music-model-kit/lib/midi.js
const log = require("@jasonfleischer/log")

class Midi {
	
	constructor(noteOn, noteOff){

		this.noteOn = noteOn;
		this.noteOff = noteOff;
		this.devices = [];

		let self = this

		if (!navigator.requestMIDIAccess) {
			log.e("this browser does not support midi");
			return
		}

		navigator.permissions.query({ name: 'midi-sysex'}).then(function(result) {
			if(result.state == 'granted') {
				log.e("Midi permissions granted");
			} else {
				log.e("Midi permissions denied");
			}
		});


		navigator.requestMIDIAccess().then(function(access) {
			if (access.inputs.size > 0) {
				self.devices = access.inputs.values();
				self.connectToFirstDevice();
			}
			access.onstatechange = function(e) {
				self.devices = Array.from(this.inputs.values());
				self.connectToFirstDevice();
			}
    	}, function() {
    		log.e("Midi Falure");
    	});
	}

	connectToFirstDevice() {
		if (this.devices > 0) {
			let device = this.devices[0];
			this.connectToDevice(device)
		}
	}

	connectToDevice(device) {
	  	log.e('Connecting to device: ' + device);
	  	this.connectedDevice = device;
	  	device.onmidimessage = function(m) {
		    const [command, midi_value, velocity] = m.data;

		    let KEY_UP = 145;
		    let KEY_DOWN = 129;

		    if (command === KEY_UP) {
		    	log.e('KEY UP: ' + midi_value);
		    	this.noteOff(midi_value, velocity);
		    } else if(command === KEY_DOWN) {
		      	log.e('KEY DOWN'+ midi_value);
		      	this.noteOn(midi_value, velocity);
		    }
		 }
	}


	deviceToString(device) {
		return device.name + ' ' + device.manufacturer;
	}
}

module.exports = Midi
