// @jasonfleischer/music-model-kit/lib/midi.js
const log = require("@jasonfleischer/log")

class Midi {
	
	constructor(noteOn, noteOff){

		this.noteOn = noteOn;
		this.noteOff = noteOff;
		this.devices = [];

		let self = this

		let devices = this.devices;

		if (!navigator.requestMIDIAccess) {
			log.e("this browser does not support midi");
			return
		}

		/*navigator.permissions.query({name: 'midi-sysex'}).then(function(result) {
			if(result.state == 'granted') {
				log.e("Midi permissions granted");
			} else if (result.state == 'prompt'){
				log.e("Midi permissions prompt");
			} else {
				log.e("Midi permissions denied");
			}
		});

		navigator.permissions.query({name: 'sysex'}).then(function(result) {
			if(result.state == 'granted') {
				log.e("Midi permissions granted");
			} else if (result.state == 'prompt'){
				log.e("Midi permissions prompt");
			} else {
				log.e("Midi permissions denied");
			}
		});

		navigator.permissions.query({name: 'midi'}).then(function(result) {
			if(result.state == 'granted') {
				log.e("Midi permissions granted!!");
			} else if (result.state == 'prompt'){
				log.e("Midi permissions prompt");
			} else {
				log.e("Midi permissions denied");
			}
		});*/

		navigator.permissions.query({name: 'midi', sysex: true}).then(function(result) {
			if(result.state == 'granted') {
				log.e("Midi permissions granted!");


/*navigator.requestMIDIAccess({sysex: true}).then(function(access) {
			if (access.inputs.size > 0) {
				log.e("9")
				devices = access.inputs.values();
				self.connectToFirstDevice();
			} else {
				log.e("no midi inputs " + access.inputs.size);
			}
			access.onstatechange = function(e) {
				log.e("Midi State change");
				devices = Array.from(this.inputs.values());
				self.connectToFirstDevice();
			}
    	}, function() {
    		log.e("Midi Falure");
    	});*/



			} else if (result.state == 'prompt'){
				log.e("Midi permissions prompt");
			} else {
				log.e("Midi permissions denied");
			}
		});

		navigator.requestMIDIAccess({sysex: true}).then(function(access) {
			if (access.inputs.size > 0) {
				log.i("Midi input size: " + access.inputs.size)
				self.devices = access.inputs.values();
				self.connectToFirstDevice(Array.from(access.inputs.values()));
			} else {
				log.e("no midi inputs 2: "  + access.inputs.size);
			}
			access.onstatechange = function(e) {
				log.e("Midi State change 2");
				self.devices = Array.from(this.inputs.values());
				self.connectToFirstDevice(Array.from(access.inputs.values()));
			}
    	}, function() {
    		log.e("Midi Falure");
    	});
	}

	connectToFirstDevice(devices) {
		log.e("Len: "  + devices.length);
		if (devices.length > 0) {
			let device = devices[0];
			this.connectToDevice(device);
		} else {
			log.e("connectToFirstDevice: no midi inputs");	
		}
	}

	connectToDevice(device) {
	  	log.e('Connecting to device: ' + device);
	  	this.connectedDevice = device;

	  	let noteOn = this.noteOn;
		let noteOff = this.noteOff;
		let NOTE_ON = 0x9;
		let NOTE_OFF = 0x8;
	  	device.onmidimessage = function(m) {
		    const [command, message, velocity] = m.data;


		    //log.e("Input" +command + "," + message + ', ', velocity);
		    let midi_value = message;
		    
		    let channel = command & 0x0F;
		    let opCode = (command & 0xF0) >> 4;
 
		    if(midi_value == 60){
		    	log.e("Input" +command + "," + message + ', ', velocity);
		    	log.e(opCode + " V" + midi_value + " ch" + channel);
			}

		    if (opCode === NOTE_ON) {
		    	log.e('NOTE_ON: ' + midi_value);
		    	noteOff(midi_value, channel, velocity);
		    } else if(command === NOTE_OFF) {
		      	log.e('NOTE_OFF: '+ midi_value);
		      	noteOn(midi_value, channel, velocity);
		    }
		 }
	}


	deviceToString(device) {
		return device.name + ' ' + device.manufacturer;
	}
}

module.exports = Midi
