// @jasonfleischer/piano/index.js
const musicKit = require("@jasonfleischer/music-model-kit");
const PianoView = require("./lib/piano_view.js");
const log = require("@jasonfleischer/log")


class PianoBuilder {

	constructor(options) {

		this.id = options.id;
		if (id === undefined){
			log.e('id not provided for piano')
			return
		}

		this.pianoView = document.getElementById(id);
		if (pianoView === undefined){
			log.e('not piano exists with id: ' + id)
			return
		}

		function isInt(value) {
			var x = parseFloat(value);
			return !isNaN(value) && (x | 0) === x;
		}
		this.range = musicKit.piano_range;
		if (options.range !== undefined){
			if (options.range.min !== undefined && options.range.max !== undefined) {
				if(isInt(options.range.min)){
					this.range.min = Math.min(Math.max(range.min, options.range.min), range.max);
				}
				if(isInt(options.range.max)){
					this.range.max = Math.min(Math.max(range.min, options.range.max), range.max);
				}
			}
		}

		this.width = 1000;
		if (options.width !== undefined){
			this.width = options.width
		}

		this. interactive = false;
		if (options.interactive !== undefined){
			this.interactive = options.interactive;
		}

		
		log.e("?" + this.range.min +' ' + this.range.max );   
		this.view = new PianoView(this.id, this.width, this.range, this.interactive);
		return this.view;
	}
}


module.exports = { PianoBuilder };
